### ISSUE TITLE
Substitution procedure (2019 rules)

### ISSUE THREAD

**Issue by BernardoCunha** [RoboCup-MSL/RefBox2015#31]
The 2019 rules introduced the possibility of doing fully autonomous substitutions of robots. For this action, the number of the robot on the field that is supposed to be replaced must be send by the RefBox together with a "substitute" command. First of all, the protocol to be used must be defined since it does not exist yet.

**Comment by jorritolthuis** [comment]
Rules RC-3.5 describe the substitution procedure. Unfortunately, this leaves many implementation details open.
I have created a chart to indicate my view on the procedure. It shows a the process of a request that one team (team A) asks to substitute robot 3 with robot 4:

![RefBox-substitution](https://user-images.githubusercontent.com/28347783/55557262-c5f44700-56e9-11e9-8ebb-17d1eea2f6b8.png)

I am aware of the fact that many different implementations are possible (with more or less done by the teams). I have tried to choose a middle ground for this. If there are any good reasons to take another approach, please let me know.

The general idea is that the RefBox referee will select two robots for the procedure. During the whole process, the general game state will be “stop”. A unique signal (indicating a specific robot) is sent to the base station (only of the switching team). This robot has to leave the field.
Next, another unique signal (from another set, indicating another robot) is sent to the base station. At that point, the substituting robot can drive to the sideline.
A design decision here is that the RefBox referee will indicate when the substituting robot can start, instead of having teams determine this moment.

There are a few comments I need to make for clarification:
* For leaving the field, the rules specify 15 seconds. For the complete switch, 20 seconds. I propose to set a timer (sound) only for the 15 seconds since this is the least ambiguous to measure.
* The rules specify a signal to certain robots. Since there is no connection between the RefBox and robots, this signal is going to the team’s base station and must be handled there.
* Since I was not able to find any RoboCup change to the number of nominated substitutes, teams should have 3 up to 7 nominated substitutes. Since I don’t think any team has this, I assumes a total of 6 players, including nominated substitutes.
* Many alphanumeric characters have already been used in the communication protocol. Since I value subsequent characters for subsequent robot numbers, I have chosen to use a character set. (The way this has been chosen, allows to easily increase the 6 of the previous point to 7).

**Comment by BernardoCunha** [comment]
Dear Jorrit

Thanks for your insights in the substitution procedure. In fact, as previously discussed among the TC and EXECs, this procedure does not only require changes to the current  RefBox but also to the current used protocol. The single letter approach that we use does no longer support a procedure like this (according to the rules, robots can have numbers with one or two digits and this implies that a basic change to the protocol must be implemented). The question, at this point in time, is that it is simply to late to bring out a new protocol to the teams and enforce that protocol for the current year. That‘s why I already suggested to the TC, by the end of January, that a new protocol should be brought to MSL discussion, so that we can agree during 2019 on a new solution to be applied in 2020.

That said, the way I see the replacement procedure is simpler than the one you describe. This has also been debated among the TC and an agreement has been reached already. So:
- The team that want to perform a substitution should ask the main referee and the Rbox referee for a substitution, and indicate the number of the robot that is going to be replaced. 
- This must be done using a visible cardboard with the number of the robot that is going to be substituted. Of course this can only happen during a game stoppage.
- The RefBox will then send a command indicating the number of the robot that has to leave the field.
- According to the current discussion, the “cyan/magenta” protocol strategy will also be replaced by the concept “own/opponent” team concept.
- In this case, the team that will replace the robot will receive a command in the form “own/replace/robot_number” while the other team will receive a command in the form “opponent/replace/robot_number”
- For the robot that is going to replace the robot that is leaving the field there is no action from the RefBox. This robot must simply be activated inside the repair zone, and should enter the field and stay close to the outside line as soon as its team mate has entered the repair zone.

So, basically, according to the discussion among the TC, the repair procedure will not not be used in 2019. Contributions or proposals for the new protocol and help on changing the RefBox are, however, very welcome.

By the way. Changes to the RefBox for the new repair times and procedures are already integrated in a new version that will be released soon.

Best regards

Bernardo Cunha

**Comment by jorritolthuis** [comment]
Dear Bernardo,

I am aware that there is too little time available to implement and thoroughly test this protocol before RoboCup 2019 (and give teams enough time to adapt).  Moreover, Wouter Houtman had already told me that the TC has the intention to start using it in 2020. Since I currently have some time to work on the RefBox, I already want to start development.

I do not agree with your statement that a new communication protocol is required for this addition. Even though the rules state that 99 numbers markings are allowed, there is no team that actually has 99 physical robots. I believe only one signal is required for every physical robot, since with that they can be identified unambiguously. 
A config file can be created to map the number markings to communicated signals (such that the RefBox shows the numbers of the marking, but sends a character of a much smaller set). The base station can then locally translate the character back to a physical robot (since the teams know which robot carries which label according to the RefBox). This way, the referee can use the numbers on the markings, and there is no need to revise the protocol.

Nonetheless, I agree that it might be good to revise some parts of the RefBox (e.g. communication, own/opponent instead of cyan/magenta, refactor code). This will make sure the RefBox can be updated/extended properly in the future. However, I think it would be a shame if the substitution procedure has to wait on that. I doubt whether all changes will be finished and tested in time to use it during RoboCup 2020.

I did not include any communication to the opponent team, since I think there is little relevance for them to know about this substitution of the other team. If you think this is relevant, it is of course possible to include a character to inform the opponent of a substitution.

> For the robot that is going to replace the robot that is leaving the field there is no action from the RefBox. This robot must simply be activated inside the repair zone, and should enter the field and stay close to the outside line as soon as its team mate has entered the repair zone.

I did not get this from reading the rules, but this would simplify the process on the RefBox side significantly. I have updated my previous figure to show the process of replacing robot 25 by robot 4:

![RefBox-substitution2](https://user-images.githubusercontent.com/28347783/55630651-79c30880-57b6-11e9-83e6-8cd65367c28c.png)

**Comment by BernardoCunha** [comment]
Hi again Jorrit

See comments inline below.

All the best

Bernardo

**Comment by janfeitsma** [comment]
See the [slides](https://docs.google.com/presentation/d/1XQT6XgQZAIdV0wViBV6kJbBrObZrNrEw2TJdAbNIwJY/edit?usp=sharing) from MSL workshop 2019.

**Comment by franciscompinto** [comment]
Special case: what if the same team connects twice to RefBox resolverd.

**Comment by BernardoCunha** [comment]
Substitution procedure is now implemented in the RefBox. Since each team can make up to three substitutions during a game and the rules do not state that it must be one substitution at a time, the Pop Up window will allow up to three robot numbers from each team to be introduced.
